### PR TITLE
Specific message for string queries that do not project __id and __cas.

### LINKED ISSUE #1097: NullPointerException in @Query that do not project __cas .  Should check for this explicitly
https://forums.couchbase.com/t/unable-to-use-n1ql-query-in-query-annotation-with-max-function-in-spring-data-couchbase-4-1-5/29776

Maybe even allow it.  See could for ReactiveFindByQueryCount.all() - handling when there are distinctFields.

### PR DESCRIPTION
**Pull request by mikereiche**
Closes #1097.
Original pull request: #1114.

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATACOUCH).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

File: src/main/java/org/springframework/data/couchbase/core/ReactiveFindByQueryOperationSupport.java [hunk 1]
```diff
@@ -147,7 +148,13 @@ public ReactiveFindByQueryOperationSupport(final ReactiveCouchbaseTemplate templ
 					String id = "";
 					long cas = 0;
 					if (distinctFields == null) {
+						if( row.getString(TemplateUtils.SELECT_ID) == null){
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/ReactiveFindByQueryOperationSupport.java:151]: formatting looks off with space

File: src/main/java/org/springframework/data/couchbase/core/ReactiveFindByQueryOperationSupport.java [hunk 2]
```diff
@@ -147,7 +148,13 @@ public ReactiveFindByQueryOperationSupport(final ReactiveCouchbaseTemplate templ
...
+							throw new CouchbaseException("query did not project "+TemplateUtils.SELECT_ID+" "+statement);
+						}
 						id = row.getString(TemplateUtils.SELECT_ID);
+						if( row.getLong(TemplateUtils.SELECT_CAS) == null){
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/ReactiveFindByQueryOperationSupport.java:155]: same here

File: src/main/java/org/springframework/data/couchbase/core/ReactiveFindByQueryOperationSupport.java [hunk 3]
```diff
@@ -147,7 +148,13 @@ public ReactiveFindByQueryOperationSupport(final ReactiveCouchbaseTemplate templ
 					String id = "";
 					long cas = 0;
 					if (distinctFields == null) {
+						if( row.getString(TemplateUtils.SELECT_ID) == null){
+							throw new CouchbaseException("query did not project "+TemplateUtils.SELECT_ID+" "+statement);
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/ReactiveFindByQueryOperationSupport.java:152]: I think both exceptions need more explanation in the message for the user what to do and how to fix it

**Review by daschl — CHANGES_REQUESTED** [review]

**Comment by daschl** [comment]
also @mikereiche is this needed for analytics as well?

File: src/main/java/org/springframework/data/couchbase/core/ReactiveFindByQueryOperationSupport.java (hunk 3 above)
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/ReactiveFindByQueryOperationSupport.java]: fixed:
com.couchbase.client.core.error.CouchbaseException: query did not project __id. Either use #{#n1ql.selectEntity} or project __id and __cas : SELECT __cas, * from `b406ab45-ef95-441f-a318-76b8e8af2d76` where iata = $1

File: src/main/java/org/springframework/data/couchbase/core/ReactiveFindByQueryOperationSupport.java (hunk 2 above)
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/ReactiveFindByQueryOperationSupport.java]: fixed.

**Comment by mikereiche** [comment]
Yes, also needed in Analytics.

File: src/main/java/org/springframework/data/couchbase/core/ReactiveFindByAnalyticsOperationSupport.java [hunk 4]
```diff
@@ -92,10 +94,22 @@ public ReactiveFindByAnalyticsOperationSupport(final ReactiveCouchbaseTemplate t
...
 							}
 						}).flatMapMany(ReactiveAnalyticsResult::rowsAsObject).map(row -> {
-							String id = row.getString("__id");
-							long cas = row.getLong("__cas");
-							row.removeKey("__id");
-							row.removeKey("__cas");
+							String id = "";
+							long cas = 0;
+							if (row.getString(TemplateUtils.SELECT_ID) == null) {
+								throw new CouchbaseException("query did not project " + TemplateUtils.SELECT_ID
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/ReactiveFindByAnalyticsOperationSupport.java:100]: nit: should probably say in both exceptions "analytics query did not project".. to distinguish from a n1ql query in the logs if seen

File: src/test/java/org/springframework/data/couchbase/domain/Iata.java [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package org.springframework.data.couchbase.domain;
```
Comment (daschl) [inline src/test/java/org/springframework/data/couchbase/domain/Iata.java:1]: needs a license header?

File: src/test/java/org/springframework/data/couchbase/domain/Iata.java [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+
+public enum Iata {
+  vie,
```
Comment (daschl) [inline src/test/java/org/springframework/data/couchbase/domain/Iata.java:4]: should they be VIE and XXX since we write enums upper caps?

**Review by daschl — CHANGES_REQUESTED** [review]
minor nits, otherwise looks good to me

**Comment by mikereiche** [comment]
> minor nits, otherwise looks good to me

fixed "analytics query ..." message.
The Iata Enum needs the elements (vie) in lower-case because the (String) iata is always stored as lower-case.

**Pull request merged**